server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java

```diff
@@ -68,7 +72,20 @@ public Query rewrite(IndexSearcher searcher) throws IOException {
     @Override
     public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
         if (searcher.getIndexReader().getContext().id() != contextIdentity) {
-            throw new IllegalStateException("This KnnScoreDocQuery was created by a different reader");
+            if (searcher.getLeafContexts().size() == 1) {
+                boolean found = false;
+                LeafReader reader = FilterLeafReader.unwrap(searcher.getLeafContexts().get(0).reader());
+                for (Object id : segmentIdentites) {
+                    if (id == reader.getContext().id()) {
+                        found = true;
+                    }
+                }
+                if (found == false) {
+                    throw new IllegalStateException("This KnnScoreDocQuery was created by a different reader");
+                }
+            } else {
+                throw new IllegalStateException("This KnnScoreDocQuery was created by a different reader");
+            }
```

Comment:
This is frustrating, I am not 100% sure how to fully fix this other than this.

https://github.com/elastic/elasticsearch/blob/d411acecbcbd029ebadbfa533a07c70330d454c8/server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java#L120

So, for every hit that is within a given reader context, we set the matching docs and that individual sub-reader. This makes up the `HitContext`. This is then passed to the highlighting processor, and used when making the `FieldHighlightContext`:

https://github.com/elastic/elasticsearch/blob/28a504d7a1ee520c52a9dcac9cdee2df0deb6e14/server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/DefaultHighlighter.java#L55

Eventually, we get down to finding the offsets here in Lucene: https://github.com/apache/lucene/blob/5694e30bccd13d180cd1a813f532aa12c6ab36dc/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/FieldHighlighter.java#L80

The specific offset class being used is this one: https://github.com/apache/lucene/blob/5694e30bccd13d180cd1a813f532aa12c6ab36dc/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/MemoryIndexOffsetStrategy.java#L104C5-L104C5

Since we are using the weight_offsets, we go down this path: https://github.com/apache/lucene/blob/5694e30bccd13d180cd1a813f532aa12c6ab36dc/lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/FieldOffsetStrategy.java#L114

There we see the individual leafReader being wrapped and a new searcher being created around that reader.

So, its the SAME leaf reader, but the context we were expecting in kNN is the top level context. 

In this "fix", I:

 - Verify we just have one leaf (which indicates this path of we wrapped the individual leaf reader)
 - Unwrap it to ensure its one we used before
 - pass the test

Reply:
This will work only if we reuse the underlying index. The unified highlighter will use a memory index for cases where the highlighted fields don't have the offsets index option. I am not sure how we can fix it without disabling the weight matches mode if we detect a `KnnScoreDocQuery` internally?

Reply:
@jimczi yeah, I don't know how we can fix it either. I did disable weight matches if we detect that query kind: https://github.com/elastic/elasticsearch/pull/101713

Reply:
Ok good to know, thanks.